src/MCPServer/lib/RPCServer.py

```diff
@@ -54,206 +65,374 @@ class NotFoundError(RPCServerError):
+
+       - ``worker``: instance of the worker class which is a subclass of
+         ``gearman.job.GearmanWorker``.
+
+       - ``job``: instance of ``gearman.job.GearmanJob``.
+
+       Additionally, a method handler can have an argument named ``payload``.
+       If present, it will carry the data embedded in the payload of the job
+       already unserialized. This is useful when the handler wants to read
+       data sent by the client.
 
-    If ``raise_exc`` is True, handler exceptions are raised. This is
-    interpreted by the worker library as a failed job and the client will not
-    receive additional information about the error other than its status.
-    Set ``raise_exc`` to False when you want the client to receive a detailed
-    error.
     """
-    def decorator(func):
-        def wrap(*args, **kwargs):
+    # Regular expression to match handler configurations in method docstrings.
+    ability_regex = re.compile(
+        r"(?P<docstring>.*)(?P<config>\[config\].*)$",
+        re.DOTALL | re.MULTILINE)
+
+    # To some extent, application logic in Archivematica is coupled to certain
+    # values in our workflow data. This is mostly true to things related to
+    # types of transfers and its approvals. This value in particular supports
+    # the partial approval of AIP reingest.
+    APPROVE_AIP_REINGEST_CHAIN_ID = "9520386f-bb6d-4fb9-a6b6-5845ef39375f"
+
+    def __init__(self, workflow):
+        super(RPCServer, self).__init__(
+            host_list=[django_settings.GEARMAN_SERVER])
+        self.workflow = workflow
+        self._register_tasks()
+        self.set_client_id(gethostname() + "_MCPServer")
+
+    def _register_tasks(self):
+        for ability, handler in self._handlers():
+            logger.debug("Registering ability %s", ability)
+            self.register_task(ability, handler)
+
+    def _handlers(self):
+        members = inspect.getmembers(self, predicate=inspect.ismethod)
+        for name, handler in members:
+            if name[0] != "_":
+                continue
+            if not name.endswith("_handler"):
+                continue
+            name, opts = self._get_ability_details(handler)
+            yield name, self._wrap_handler_method(handler, **opts)
+
+    def _get_ability_details(self, handler):
+        docstring = str(inspect.getdoc(handler))
+        match = self.ability_regex.search(docstring)
+        err = ValueError("Unexpected docstring format: {}".format(docstring))
+        if match is None:
+            raise err
+        try:
+            config = match.group("config")
+        except IndexError:
+            raise err
+        parser = configparser.SafeConfigParser({
+            "name": None, "raise_exc": False,
+        })
+        parser.readfp(StringIO(config))
+        return parser.get("config", "name"), {
+            "raise_exc": parser.getboolean("config", "raise_exc"),
+            "expect_payload": "payload" in inspect.getargspec(handler).args,
+        }
```

Comment:
> Have you some more background on the pattern here, or why you chose this approach?
> Once you have a necessity for a _handler(...) function to have attributes (name, raise_exc) it starts to make them look more like objects than functions. So is there a way to leverage classes in a similar way?

Are you thinking on something like Django's [class-based views](https://docs.djangoproject.com/en/2.1/topics/class-based-views/). I've always thought they were not as simple.

When I wrote this I was just thinking on easing the transition to gRPC as much as possible, where you usually have a class and a bunch of methods that are named in advance by the service definition, e.g. see [route_guide_server.py](https://github.com/grpc/grpc/blob/eedcd2a78177b722063d924e2d401006e95d0ecd/examples/python/route_guide/route_guide_server.py#L59-L112).
My new `RPCServer` introduced application-level exception handling on top of our Gearman RPC and `raise_exc` was just about making sure that the original RPCs could still rely on the old behaviour to avoid potential regressions in this work.

> That's related to the tuple that's returned. Where classes could be used to test for attribute errors if properties aren't available i.e. they're not configured correctly.

You don't need OOP to do that. But anyways that's one of the reasons why we started talking about gRPC too, it comes with protobuf which brings typed fields to Python. Just not scoped this time. We experimented with that stuff a bit at some point, see https://github.com/artefactual/archivematica/blob/dev/issue-8161-dashboard-i18n-db/src/protos/mcpserver.proto.

> Using the [config] string, we can of course do some validation up front too, but it's worth noting that the call to getBoolean will immediately raise a ValueError if raise_exc doesn't exist or has an invalid value, whereas parser.get("config", "name") not being set will continue through the code a little longer until a gearman exception is raised.
> I think I'd like to fail as early as possible in this functionhere rather than that happening, so perhaps there's a little we can do around retrieving the name and raise_exc.

Good observation, will fix!

Reply:
I've made a couple of updates in the RPCServer to raise when the `name` attribute is missing. Also the wrapper handler had a problem in the way that exceptions were being controlled. I think errors are clearer now and handlers should return application-level errors at all times, i.e. the RPC client should receive a `{"error": True, ... }` when something wrong happened.

Reply:
Sounds good. Will have a look,